src/Pos/Generator/Block/Payload.hs

```diff
@@ -0,0 +1,235 @@
+       ( genPayload
+       ) where
+
+import           Universum
+
+import           Control.Lens               (at, uses, (%=), (.=), (?=))
+import           Control.Lens.TH            (makeLenses)
+import           Control.Monad.Random.Class (MonadRandom (..))
+import qualified Data.HashMap.Strict        as HM
+import           Data.List                  (notElem, (!!))
+import qualified Data.List.NonEmpty         as NE
+import qualified Data.Map                   as M
+import qualified Data.Vector                as V
+import           Formatting                 (sformat, (%))
+import           System.Random              (RandomGen (..))
+
+import           Pos.Client.Txp.Util        (makeAbstractTx, overrideTxDistrBoot)
+import           Pos.Context                (genesisStakeholdersM)
+import           Pos.Core                   (Address (..), Coin, SlotId (..),
+                                             addressDetailedF, coinToInteger,
+                                             makePubKeyAddress, sumCoins,
+                                             unsafeIntegerToCoin)
+import           Pos.Crypto                 (SecretKey, SignTag (SignTxIn), WithHash (..),
+                                             hash, sign, toPublic)
+import           Pos.DB                     (gsIsBootstrapEra)
+import           Pos.Generator.Block.Error  (BlockGenError (..))
+import           Pos.Generator.Block.Mode   (BlockGenRandMode, MonadBlockGenBase)
+import           Pos.Generator.Block.Param  (HasBlockGenParams (..), HasTxGenParams (..),
+                                             asSecretKeys)
+import qualified Pos.GState                 as DB
+import           Pos.Slotting.Class         (MonadSlots (getCurrentSlotBlocking))
+import           Pos.Txp.Core               (TxAux (..), TxIn (..), TxInWitness (..),
+                                             TxOut (..), TxOutAux (..), TxSigData (..))
+import           Pos.Txp.Logic              (txProcessTransaction)
+import           Pos.Txp.Toil.Class         (MonadUtxo (..), MonadUtxoRead (..))
+import           Pos.Txp.Toil.Types         (Utxo)
+import qualified Pos.Txp.Toil.Utxo          as Utxo
+
+----------------------------------------------------------------------------
+-- Tx payload generation
+----------------------------------------------------------------------------
+
+-- | Generates list of distinct ints in the given range [a,b].
+selectDistinct :: forall m. (MonadRandom m, MonadIO m) => Int -> (Int,Int) -> m [Int]
+selectDistinct n0 p@(a, b)
+    | b - a < 0 = error $ "selectDistinct: b < a " <> show p
+    | otherwise = do
+          res <- selectDistinct' [] n
+          if fromIntegral (length res) /= n
+              then error "selectDistinct is broken"
+              else pure res
+  where
+    n :: Int
+    n = min (b + 1 - a) n0
+    selectDistinct' :: [Int] -> Int -> m [Int]
+    selectDistinct' cur leftN = do
+        leftItems <- take leftN <$> getRandomRs (a, b)
```

Comment:
1. Read documentation pls.
2. It doesn't matter, list is lazy.
3. It's totally alright, i don't understand your predjawa.